### ISSUE TITLE
Compute learning curves

### ISSUE THREAD

**Issue by FrancescoCasalegno** [BlueBrain/Search#276]
## Context

- We want to be able estimate how many training samples on average are required to achieve a good accuracy for a new entity type.
- Clearly the number depends on the new entity type: if it is a completely new type the number may be higher, but if it is related to an entity for which we already have a model then the pre-trained weight could make our job easier and fewer examples may be required.
- We can use the entity types of the CORD19 dataset.

## Actions
- [x] For each entity type, plot `train_set_size` vs. `error_on_test_set`. 
   - Note: we could take several samples so that we could also confidence bars for each observation. 
   - Note: this kind of plot usually should be done in log-log.
   - Note: the y axis should really be some sort of "error" to make the log-log plot readable; e.g. for the y-axis "1 - f1 score" or "1 - recall" can be used.
   - Note: there should be one plot for each of the metrics of interest (f1, prec, recall, ...)
- [x] For each entity type, verify if there is a train set size after which the curve "saturates", i.e. the metric stops improving.

**Comment by pafonta** [comment]
Hello @FrancescoCasalegno,

Maybe we could add as a requirement that the implementation will rely on DVC plots?

Documentation: [dvc plots](https://dvc.org/doc/command-reference/plots) and [end-to-end user guide](https://dvc.org/doc/start/metrics-parameters-plots).

**Comment by pafonta** [comment]
Note: 

We might want also to use spaCy `Scorer` ([doc](https://spacy.io/api/scorer)) to compute the performance, instead of the custom `data_and_models/pipelines/ner/eval.py`.

**Comment by FrancescoCasalegno** [comment]
Sure, this new `spacy.scorer` looks nice! For what I quickly read it seems to provide a subset of the functionalities of `bluesearch.mining.eval`. 

In particular, their NER evaluation is only entity-based (we provide both token-based and entity-based evaluation, plus we also have a few advanced error analysis functions that I believe can still be quite useful.

So I would not throw away `bluesearch.mining.eval` too quickly.

**Comment by pafonta** [comment]
Sure.

We need however to discuss if it's make sense to do a token-level evaluation. Indeed, the NER component is trained and optimized for what the `Scorer` evaluates. The `Scorer` do an entity-level evaluation. This is actually written at the top of the [documentation for the NER component](https://spacy.io/api/entityrecognizer):

> the loss function optimizes for **whole entity accuracy**, so if your inter-annotator agreement on boundary tokens is low, the component will likely perform poorly on your problem

NB: Scorer was already part of spaCy 2.

**Comment by pafonta** [comment]
Regarding DVC plots, DVC 2.0.6 (latest) adds `--open` to `dvc plots show` to "opens the generated plot directly in the browser" 🔥 That could be worth to keep in mind in the context of this issue and #273. Indeed, it could prevent us from implementing something handmade for this feature.

https://dvc.org/doc/command-reference/plots/show#plots-show

**Comment by jankrepl** [comment]
> Note:
> 
> We might want also to use spaCy `Scorer` ([doc](https://spacy.io/api/scorer)) to compute the performance, instead of the custom `data_and_models/pipelines/ner/eval.py`.

We actually considered using `Scorer` before implementing `data_and_models/pipelines/ner/eval.py`, however, it seemed like a total black box to us and that is why we decided to write our own evaluation.

**Comment by FrancescoCasalegno** [comment]
# Method

The data were collected by specifying a `factor` from `0.25` to `1.00`  to each stage of the `ner` pipeline in the `params.yml`. E.g. after the lines
https://github.com/BlueBrain/Search/blob/78e82f4c0f8f04790f87424a80df622f500a9f5c/data_and_models/pipelines/ner/params.yaml#L2-L7
the line
```yml
     factor: x.xx
```
was added, where `x.xx` was set every time to `0.25`, `0.50`, `0.75`, `1.00`. Then, `dvc repro` automatically triggered the re-training and re-evaluation, so that the files in `data_and_models/metrics/ner/` were updated.

<img src="https://user-images.githubusercontent.com/17013890/112190717-8dffaa80-8c05-11eb-9af7-f1d5f7323ce1.png" width="600px">


# Comments / FAQs

1. **Why not using `prodigy`'s [`train_curve`](https://prodi.gy/docs/recipes#train-curve)?**
Because `train_curve` reports scores computed on the `valid` set, not on the `test` set. And this is a bad thing for two reasons.
   1. The `valid` set is much smaller than the `test` set, so there is much more uncertainty associated to those scores.
   2. The `test` set is a separate set of sentences from the `train` ones, while `valid` is a portion of the `train` set that is set aside. Results on the `valid` set are therefore less informative about the generalization properties of our model.
2. **Why not reporting results for various folds/splits?**
Because `prodigy`'s `train` does not allow this. Even the random seed is fixed with a `fix_random_seed(0)`.
3. **Why some scores do not improve as more training samples are used?**
   - Hard to tell, but to be kept in mind for the future, especially for #293: when we want to train a model for a new entity type, we should plot these curves to know if more training samples are required!
   - Moreoever, notice that for `PATHWAY`, which is the only entity type for which we did not have any baseline model available, the more training samples we have the better the results!
4. **Why is this `valid` set used for?** 
`prodigy train` uses the `valid` split to track the score obtained on it at the end of every epoch, and then return the model corresponding to the epoch where the best score was observed.
This sounds similar to what `sklearn` does with its model selection, but unlike `sklearn`'s [`refit`](https://github.com/scikit-learn/scikit-learn/blob/95119c13a/sklearn/model_selection/_search.py#L1063), here the returned model is fitted only on the training set, not on the train+valid set.
Moreover, for deployment in production we should consider training the final model on train+valid+test, so that we use all available data. This is in particular important when the number of training samples is less than the ones in the test set—this is the case e.g. for `PATHWAY`: 151 samples for `train`, >300 samples for `test.



# Results
![learn_curves](https://user-images.githubusercontent.com/17013890/112188998-d5853700-8c03-11eb-8a47-f3f11b98edb0.png)
